### PR TITLE
Hacker News

### PR DESCRIPTION
**Pull request by vkatsuba**
Proposed changes include:
* `doc/advanced-configuration/Modules.md` - added info about new module
* `include/mongoose_ns.hrl` - added feature name space
* `mkdocs.yml` - added new module documentation to module index
* `rel/files/mongooseim.toml` - added section
* `src/config/mongoose_config_spec.erl` - added new module to configurable modules
* `test/config_parser_SUITE.erl` - added test case for new config
* `test/config_parser_SUITE_data/modules.options` - added new options for test
* `test/config_parser_SUITE_data/modules.toml` - added config
* `big_tests/tests/mod_hacker_news_SUITE.erl` - added new test
* `big_tests/default.spec` - added new spec for new test
* `doc/modules/mod_hacker_news.md` - added documentation
* `src/mod_hacker_news.erl` - added new module

**Tests Result**: https://github.com/vkatsuba/MongooseIM/runs/2615143399?check_suite_focus=true

An application that does the same - https://github.com/vkatsuba/hacker_news

---

**Request:**
```xml
<iq type='get' id='hacker_news_get'>
	<query xmlns='erlang-solutions.com:xmpp:hacker-news'/>
</iq>
```
**Response:**
```xml
<iq from='tester@localhost' to='tester@localhost/66aa4d329a4872851621-587277-755414' id='hacker_news_get' type='result'>
	<items xmlns='erlang-solutions.com:xmpp:hacker-news'>
		<topstories>[27230427,27231471,27230774,27226382,27230042,27229836,27224094,27193099,27227683,27210022,27220787,27230190,27222457,27226625,27213882,27207604,27229436,27223340,27229171,27225789,27209086,27227751,27207610,27191329,27220657,27205586,27227180,27216924,27206055,27221552,27228550,27223520,27223012,27226751,27222985,27219205,27222460,27227492,27224691,27231463,27225739,27216376,27230422,27207734,27217442,27222471,27219646,27223090,27191360,27212127]</topstories>
	</items>
</iq>
```
**API:**
```erlang
(mongooseim@localhost)1> mod_hacker_news:get_topstories().
[27230427,27231471,27230774,27226382,27230042,27229836,
 27193099,27227683,27210022,27230190,27220787,27222457,
 27226625,27207604,27213882,27229436,27216457,27229171,
 27225789,27227751,27207610,27220657,27205586,27216924,
 27206055,27227180,27228550,27224094,27226751|...]
 ```

### CODE REVIEW

**Comment by vkatsuba** [comment]
CC @chrzaszcz

File: src/mod_hacker_news.erl [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+cache_topstories(#{url := Url, total := Total,
+                   interval := Interval,
+                   retry := Retry, domain := Domain} = Opts) ->
+    _ = mongoose_metrics:update(Domain, [?MODULE, ?METRIC_TOPSTORIES], 1),
+    _ = timer:apply_after(Interval, ?MODULE, ?FUNCTION_NAME, [Opts]),
+    case maybe_retry(Url, Retry) of
+        error ->
+            ok;
+        {ok, Res} ->
+            ok = persistent_term:put(?CACHE_TOPSTORIES, lists:sublist(Res, Total))
```
Comment (chrzaszcz) [inline src/mod_hacker_news.erl:140]: The only issue here is that the module cannot be configured differently for different hosts as there is only one common storage.

File: src/mod_hacker_news.erl [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+-spec get_topstories() -> Res :: list().
+get_topstories() ->
+    persistent_term:get(?CACHE_TOPSTORIES, []).
+
+-spec cache_topstories(Data :: maps:map()) -> ok.
+cache_topstories(#{url := Url, total := Total,
+                   interval := Interval,
+                   retry := Retry, domain := Domain} = Opts) ->
+    _ = mongoose_metrics:update(Domain, [?MODULE, ?METRIC_TOPSTORIES], 1),
+    _ = timer:apply_after(Interval, ?MODULE, ?FUNCTION_NAME, [Opts]),
```
Comment (chrzaszcz) [inline src/mod_hacker_news.erl:135]: I don't see a benefit in comparing to `_`.

File: src/mod_hacker_news.erl [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+-spec maybe_retry(Url :: list(), Retry :: integer()) -> {ok, Res :: list()} | error.
+maybe_retry(_, 0) ->
+    error;
+maybe_retry(Url, Retry) ->
+    try 
+        {ok, {{_, 200, _}, _, Body}} = httpc:request(Url),
+        Res = jiffy:decode(Body),
+        {ok, Res}
+    catch
+        _:_ ->
```
Comment (chrzaszcz) [inline src/mod_hacker_news.erl:152]: It might be good to log the nature of the error, especially when we run out of retries.

**Review by chrzaszcz — COMMENTED** [review]
The implementation looks good to me. The only major issue is the single point of data storage which conflicts with the ability to configure different settings for each host.

File: src/mod_hacker_news.erl (hunk 1 above)
Comment (vkatsuba) [reply src/mod_hacker_news.erl:140]: Yep, good point, I suppose for fix it, we can add host as postfix or prefix of `CACHE_TOPSTORIES` eg:

```erlang
-define(CACHE_TOPSTORIES(Host), list_to_atom(binary_to_list(<<Host/binary, "_cache_topstories">>))).
...
CacheName = CACHE_TOPSTORIES(Host),
ok = persistent_term:put(CacheName, lists:sublist(Res, Total))
...
```
this way each host will have its own cache.

File: src/mod_hacker_news.erl (hunk 2 above)
Comment (vkatsuba) [reply src/mod_hacker_news.erl:135]: Yep, this is only trying keep one code style, only for cosmetic vision

File: src/mod_hacker_news.erl (hunk 3 above)
Comment (vkatsuba) [reply src/mod_hacker_news.erl:152]: True, totally agree.

**Comment by vkatsuba** [comment]
> The implementation looks good to me. The only major issue is the single point of data storage which conflicts with the ability to configure different settings for each host.

Agree, I'm missed this moment...

**Comment by vkatsuba** [comment]
Closed. This was only for test.

**Pull request closed without merging**